### ISSUE TITLE
Address the feedback on ABCI++ spec gathered during Informal's Protocol Design review sessions

### ISSUE THREAD

**Issue by sergio-mena** [tendermint/tendermint#7667]
## Summary

Informal's Protocol Design team is holding several sessions to go through the ABCI++ specification.

## Problem Definition

Some points to be handled have already been made. More are expected in the next few days.
We need to have the ABCI++ spec adapted to those points raised.

## Dependencies

- Wait for the end of the discussions (next week)

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

**Comment by sergio-mena** [comment]
Feedback from @mpoke:

Basic concepts and definitions
• It jumps directly into Connections / Errors … without an overview; this makes hard to follow, especially the function names.
• There is no ABCI vs ABCI++ overview.
• First describe a concept and then add examples of where it is used. For example, in the Events section, you start to describe different methods that were not yet introduced. 
• Last sentence in VoteExtensions is unclear.
• In general, if this is the entry point, then it should contain a bit more context for the reader (some overview, diagram, high-level description).  

Application requirements
• Equations are not visible on github (i.e., $p_i$)
• The structure is not clear, i.e., bullet points vs paragraphs. This makes it hard to find something without reading the entire section. Also, some statements may be lost (e.g., third sentence is not really emphasized and I assume it’s important)
• In Requirement 4, you redefine p and q, which where defined in the beginning of the section.
• In Requirement 7, $e^r_p$ is not defined

Tendermint’s expected behavior
• When describing the grammar line by line, add a comment with the path to get there, e.g., start -> clean-start / recovery -> consensus-exec before describing consensus-exec
• Not clear why a non-proposer starts with calling VerifyVoteExtension multiple times.
• “We have kept some of the ABCI++ methods out of the grammar” here should be ABCI
• I’d add the functions in the last section (i.e., Adapting existing Applications that use ABCI) as pseudocode.

Methods
• Some methods are missing (e.g., BeginBlock)
• Why is the Field Number relevant to the spec?
• For some methods, the Usage section is very large (e.g., PrepareProposal). I’d try to add some structure, e.g., pre- and post-conditions, pseudocode.  
• For ProcessProposal, under When does Tendermint call it?
   ◦ You state that the validator p may be the same as the proposer q, and then at #5 you state that “If Tendermint should prevote for the block just received” then it calls ProcessProposal. This is kind of unclear. It depends on the condition to prevote. For example, if a proposer prevotes on its own proposals, this contradicts the statement from Tendermint’s Expected Behavior “ProcessProposal will be called exactly once at all processes except the proposer”.
   ◦ # 2 -> PrepareProposal has only 5 steps.

**Comment by josef-widder** [comment]
@angbrav, I like to proposal regarding restructuring "Basic concepts". Could you please go ahead and open a PR that has the structure you propose?

**Comment by sergio-mena** [comment]
Closed by #8796 (and its backport #8936)
